crates/chain-config/src/config/state_importer.rs

```diff
@@ -0,0 +1,311 @@
+        &self.element_cursors
+    }
+}
+
+impl Read for InMemorySource {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.data.read(buf)
+    }
+}
+
+impl Seek for InMemorySource {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> std::io::Result<u64> {
+        self.data.seek(pos)
+    }
+}
+
+// So that we may record how much was written before we give it to `source`
+struct TrackingBuffReader<T> {
+    amount_read: u64,
+    source: BufReader<T>,
+}
+
+impl<T: Seek> Seek for TrackingBuffReader<T> {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> std::io::Result<u64> {
+        self.source.seek(pos)
+    }
+}
+
+impl<T: Read> TrackingBuffReader<T> {
+    pub fn new(source: T) -> Self {
+        Self {
+            amount_read: 0,
+            source: BufReader::new(source),
+        }
+    }
+
+    // unfortunately this is the best way i can find to check if there is any more data. the actual
+    // `has_data_left` method is yet to be stabilized
+    pub fn has_data_left(&mut self) -> std::io::Result<bool> {
+        Ok(!self.source.fill_buf()?.is_empty())
+    }
+}
+
+impl<T: Read> Read for TrackingBuffReader<T> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        let amount = self.source.read(buf)?;
+        self.amount_read += amount as u64;
+        Ok(amount)
+    }
+}
+
+pub struct StateReader<R> {
+    source: TrackingBuffReader<R>,
+}
+
+impl<R: Read + Seek> StateReader<R> {
+    pub fn new(source: R, start_cursor: u64) -> std::io::Result<Self> {
+        let mut reader = TrackingBuffReader::new(source);
+        reader.seek(std::io::SeekFrom::Start(start_cursor))?;
+        Ok(Self { source: reader })
+    }
+
+    pub fn batch_cursor(&self) -> u64 {
+        self.source.amount_read
+    }
+
+    pub fn read_batch<T: DeserializeOwned>(&mut self) -> anyhow::Result<Vec<T>> {
+        let coins = if self.source.has_data_left()? {
+            bincode::serde::decode_from_std_read(
+                &mut self.source,
+                Configuration::<LittleEndian, Varint, NoLimit>::default(),
+            )?
+        } else {
+            vec![]
+        };
+
+        Ok(coins)
+    }
+}
+
+struct StateWriter<W: Write> {
+    dest: BufWriter<W>,
+}
+
+use std::fmt::Debug;
+impl<W: Write + Debug> StateWriter<W> {
+    pub fn new(dest: W) -> Self {
+        Self {
+            dest: BufWriter::new(dest),
+        }
+    }
+
+    pub fn write_batch(&mut self, coins: Vec<impl serde::Serialize>) -> anyhow::Result<()> {
+        bincode::serde::encode_into_std_write(
+            coins,
+            &mut self.dest,
+            Configuration::<LittleEndian, Varint, NoLimit>::default(),
+        )?;
+
+        Ok(())
+    }
+
+    pub fn flush(&mut self) -> std::io::Result<()> {
+        self.dest.flush()
+    }
+
+    pub fn into_inner(self) -> Result<W, IntoInnerError<BufWriter<W>>> {
+        self.dest.into_inner()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct StateImporter<T> {
```

Comment:
Would it be possible to specify what kind of state, rather than just general `StateImporter`? Like `ChainState` or whatever is most appropriate.